doc/content/xcp-networkd/host-network-device-ordering.md

```diff
@@ -0,0 +1,181 @@
+device assigned a unique position in the order as its identification. Network
+devices with the same position will connect to the same network.
+
+
+The assumption
+--------------
+
+Why can the Toolstack components on two hosts independently work out an expected
+order without any communication? This is possible only under the assumption that
+the hosts have identical hardware, firmware, software, and the way
+network devices are plugged into them. For example, an administrator will always
+plug the network devices into the same PCI slot position on multiple hosts if
+they want these network devices to connect to the same network.
+
+The ordering is considered consistent if the positions of such network devices
+(plugged into the same PCI slot position) in the generated orders are the same.
+
+Particularly, when this assumption holds, a consistent initial order can be
+worked out on multiple hosts independently with the help of `biosdevname`. The
+"all_ethN" policy of the `biosdevname` utility can generate a device order based
+on whether the device is embedded or not, PCI cards in ascending slot order, and
+ports in ascending PCI bus/device/function order breadth-first. Since the hosts
+are identical, the orders generated by the `biosdevname` are consistent across
+the hosts.
+
+
+The initial order
+-----------------
+
+As mentioned above, the `biosdevname` can be used to generate consistent orders
+for the network devices on multiple hosts. However, this only works if the
+assumption of identical hardware, firmware, software, and network device
+placement holds. When this assumption doesn't hold, the order can be generated
+from a user's configuration. However, administrators would prefer to avoid this
+as much as possible when managing many hosts.
+
+
+Keep the order as stable as possible
+------------------------------------
+
+Once an initial order is created on an individual host, it should be kept as
+stable as possible across host boot-ups and at runtime. For example, unless
+there are hardware changes, the position of a network device in the initial
+order should remain the same regardless of how many times the host is rebooted.
+
+To achieve this, the initial order should be saved persistently on the host's
+local storage so it can be referenced in subsequent orderings. When performing
+another ordering after the initial order has been saved, the position of a
+currently unordered network device should be determined by finding its position
+in the last saved order. The MAC address of the network device is a reliable
+attribute for this purpose, as it is considered unique for each network device
+globally.
+
+Therefore, the network devices in the saved order should have their MAC
+addresses saved together, effectively mapping each position to a MAC address.
+When performing an ordering, the stable position can be found by searching the
+last saved order using the MAC address.
```

Comment:
Tying mac and ethX is plausible but contradicts the principle that PCI slot order (which is observable from the outside) determines NIC/ethX assignment.

Reply:
One example is:
initially, `a:AA:0 c:CC:1 b:DD:2`,  where the letters in lower case stands for MAC, and the ones in upper case stand for PCI address, and the nubmers are positions.
Now, a new card is plugged at BB: `a:AA x:BB c:CC b:DD`
The PCI slot order will result the `x:BB` with a position `1`. But what is expected is `a:AA:0 x:BB:3 c:CC:1 b:DD:2`.